uitools/cpp/Esri/ArcGISRuntime/Toolkit/FloorFilterController.cpp

```diff
@@ -0,0 +1,698 @@
+
+  bool FloorFilterController::isSelectedSiteRespected() const
+  {
+    return m_selectedSiteRespected;
+  }
+
+  void FloorFilterController::setIsSelectedSiteRespected(bool isSelectedSiteRespected)
+  {
+    if (isSelectedSiteRespected == m_selectedSiteRespected)
+      return;
+
+    m_selectedSiteRespected = isSelectedSiteRespected;
+    emit isSelectedSiteRespectedChanged();
+  }
+
+  FloorFilterController::UpdateLevelsMode FloorFilterController::updateLevelsMode() const
+  {
+    return m_updatelevelMode;
+  }
+
+  void FloorFilterController::setUpdateLevelsMode(UpdateLevelsMode updateLevelsMode)
+  {
+    if (updateLevelsMode == m_updatelevelMode)
+      return;
+
+    m_updatelevelMode = updateLevelsMode;
+    emit updateLevelsModeChanged();
+  }
+
+  FloorFilterController::AutomaticSelectionMode FloorFilterController::automaticSelectionMode() const
+  {
+    return m_automaticSelectionMode;
+  }
+
+  void FloorFilterController::setAutomaticSelectionMode(AutomaticSelectionMode automaticSelectionMode)
+  {
+    if (m_automaticSelectionMode == automaticSelectionMode)
+      return;
+
+    m_automaticSelectionMode = automaticSelectionMode;
+    emit automaticSelectionModeChanged();
+  }
+
+  void FloorFilterController::tryUpdateSelection()
+  {
+    if (m_settingViewpoint)
+      return;
+
+    Viewpoint observedViewpoint;
+    if (auto sceneView = qobject_cast<SceneViewToolkit*>(m_geoView))
+    {
+      observedViewpoint = sceneView->currentViewpoint(ViewpointType::CenterAndScale);
+    }
+    else if (auto mapView = qobject_cast<MapViewToolkit*>(m_geoView))
+    {
+      observedViewpoint = mapView->currentViewpoint(ViewpointType::CenterAndScale);
+    }
+
+    // Expectation: viewpoint is center and scale
+    if (m_automaticSelectionMode == AutomaticSelectionMode::Never ||
+        observedViewpoint.isEmpty() ||
+        isnan(observedViewpoint.targetScale()))
+      return;
+
+    auto floorManager = getFloorManager(m_geoView);
+
+    // Only take action if viewpoint is within minimum scale. Default minscale is 4300 or less (~zoom level 17 or greater)
+    double targetScale = 0.0;
+    if (floorManager)
+    {
+      if (auto siteLayer = floorManager->siteLayer())
+      {
+        targetScale = siteLayer->minScale();
+      }
+    }
+
+    if (targetScale == 0.0)
+      targetScale = 4300.0;
+
+    if (observedViewpoint.targetScale() > targetScale)
+    {
+      if (m_automaticSelectionMode == AutomaticSelectionMode::Always)
+      {
+        setSelectedLevelId({});
+        setSelectedFacilityId({});
+        setSelectedSiteId({});
+      }
+      // Assumption: if too zoomed out to see sites, also too zoomed out to see facilities
+      return;
+    }
+
+    if (!floorManager)
+      return;
+
+    // If the centerpoint is within a site's geometry, select that site.
+    // This code gracefully skips selection if there are no sites or no matching sites
+    auto sites = floorManager ? floorManager->sites() : QList<FloorSite*>{};
```

Comment:
Don't need to check the validity of floorManager because it was checked on the line before.
```suggestion
    auto sites = floorManager->sites()
```